airbrakes/data_handling/data_processor.py

```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
+            or packet.estCompensatedAccelY is None
+            or packet.estCompensatedAccelZ is None
+            for packet in self._data_packets
         ):
-            # Accelerations are in m/s^2
-            x_accel = data_packet.estCompensatedAccelX
-            y_accel = data_packet.estCompensatedAccelY
-            z_accel = data_packet.estCompensatedAccelZ
-            # Angular rates are in rads/s
-            gyro_x = data_packet.estAngularRateX
-            gyro_y = data_packet.estAngularRateY
-            gyro_z = data_packet.estAngularRateZ
-
-            # Check for missing data points
-            if any(val is None for val in [x_accel, y_accel, z_accel, gyro_x, gyro_y, gyro_z]):
-                return rotated_accelerations
-
-            # scipy docs for more info: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html
-            # Calculate the delta quaternion from the angular rates
-            delta_rotation = R.from_rotvec([gyro_x * dt, gyro_y * dt, gyro_z * dt])
-
-            # Update the current orientation by applying the delta rotation
-            current_orientation = current_orientation * delta_rotation
-
-            # Rotate the acceleration vector using the updated orientation
-            rotated_accel = current_orientation.apply([x_accel, y_accel, z_accel])
-            # Vertical acceleration will always be the 3rd element of the rotated vector,
-            # regardless of orientation. For simplicity, we multiply by -1 so that acceleration
-            # during motor burn is positive, and acceleration due to drag force during coast phase
-            # is negative.
-            rotated_accelerations[idx] = -rotated_accel[2]
+            return rotated_accelerations
+
+        # get the compensated accelerations from each data packet
+        compensated_accelerations = np.array(
+            [
+                [
+                    packet.estCompensatedAccelX,
+                    packet.estCompensatedAccelY,
+                    packet.estCompensatedAccelZ,
+                ]
+                for packet in self._data_packets
+            ]
+        )
+
+        # check if the quaternions are too uncertain or not. If they are not, we calculate
+        # the rotated accelerations directly with the quaternion data, if they are uncertain,
+        # we calculate the quaternions from the gyroscope data, and then find the rotations.
+        last_packet = self._data_packets[-1]
+        quaternion_uncertainties = np.array(
+            [
+                last_packet.estAttitudeUncertQuaternionW,
+                last_packet.estAttitudeUncertQuaternionX,
+                last_packet.estAttitudeUncertQuaternionY,
+                last_packet.estAttitudeUncertQuaternionZ,
+            ]
+        )
+
+        if (
+            np.any(quaternion_uncertainties >= QUATERNION_UNCERTAINTY_THRESHOLD)
+            or self._data_packets[-1].estOrientQuaternionW is None
```

Comment:
why is only the W quaternion checked for None? To account for missing quaternions in interest launch data? Then imo we should modify that dataset also, to add a random value, maybe exactly 0.0. We should also add `invalid_fields` for that dataset.

Reply:
yeah I'll go back and flesh it out a bit better. You think I should just check for missing quaternions and missing quaternion uncertainties only? Don't really know if checking for missing accel/gyro is worth it

Reply:
I don't think we should check any of that, they are always guaranteed to be there. Let's just modify the older datasets